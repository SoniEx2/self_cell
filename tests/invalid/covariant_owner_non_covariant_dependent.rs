use std::{cell::Cell, marker::PhantomData};

use self_cell::self_cell;

self_cell! {
    struct Foo<'a> {
        owner: PhantomData<&'a ()>,
        #[not_covariant]
        dependent: Dependent,
    }
}

type Dependent<'q> = Cell<&'q str>;

fn main() {
    let foo = Foo::new(PhantomData, |_| Cell::new(""));
    let s: String = "Hello World".into();
    foo.with_dependent(|_, d| {
        d.set(&s);
    });
    drop(s);
    foo.with_dependent(|_, d| println!("{}", d.get()));
}
