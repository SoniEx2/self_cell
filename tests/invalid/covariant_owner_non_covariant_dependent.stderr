error[E0597]: `s` does not live long enough
  --> $DIR/covariant_owner_non_covariant_dependent.rs:19:16
   |
18 |     foo.with_dependent(|_, d| {
   |                        ------ value captured here
19 |         d.set(&s);
   |                ^ borrowed value does not live long enough
...
23 | }
   | -
   | |
   | `s` dropped here while still borrowed
   | borrow might be used here, when `foo` is dropped and runs the `Drop` code for type `Foo`
   |
   = note: values in a scope are dropped in the opposite order they are defined

error[E0505]: cannot move out of `s` because it is borrowed
  --> $DIR/covariant_owner_non_covariant_dependent.rs:21:10
   |
18 |     foo.with_dependent(|_, d| {
   |                        ------ borrow of `s` occurs here
19 |         d.set(&s);
   |                - borrow occurs due to use in closure
20 |     });
21 |     drop(s);
   |          ^ move out of `s` occurs here
22 |     foo.with_dependent(|_, d| println!("{}", d.get()));
   |     --- borrow later used here
